File: backend/app/schemas/analytics.py
from __future__ import annotations

from datetime import datetime
from pydantic import BaseModel, ConfigDict, Field


class AnalyticsEventCreate(BaseModel):
    event_type: str = Field(..., description="Type of study or interaction event")
    entity_type: str = Field("roadmap_node", description="Entity type: roadmap_node, document, session, quiz, model, note")
    entity_id: str | None = Field(None, description="Optional ID of associated entity")
    value: float | None = Field(None, description="Numeric payload such as study minutes, score, latency")
    metadata_json: str | None = Field(None, description="Optional JSON string with event context")


class AnalyticsEventResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    profile_id: str
    event_type: str
    entity_type: str
    entity_id: str | None
    value: float | None
    metadata_json: str | None
    occurred_at: datetime


class ChapterProgressItem(BaseModel):
    chapter_title: str
    total_topics: int = 0
    completed_topics: int = 0
    completion_percentage: float = 0.0
    average_mastery: float = 0.0


class StudyTimeBreakdown(BaseModel):
    roadmap_minutes: int = 0
    chat_minutes: int = 0
    quiz_minutes: int = 0
    notes_minutes: int = 0
    total_minutes: int = 0


class AnalyticsOverviewResponse(BaseModel):
    completion_percentage: float = Field(0.0, description="Overall roadmap completion percentage")
    total_topics: int = 0
    completed_topics: int = 0
    in_progress_topics: int = 0
    not_started_topics: int = 0
    average_mastery: float = Field(0.0, description="Average mastery score from 0.0 to 1.0")
    total_study_minutes: int = 0
    active_streak_days: int = 0
    total_sessions: int = 0
    total_documents: int = 0
    total_notes: int = 0
    total_quizzes_taken: int = 0
    average_quiz_score: float = 0.0
    study_time_breakdown: StudyTimeBreakdown = Field(default_factory=StudyTimeBreakdown)
    chapter_progress: list[ChapterProgressItem] = Field(default_factory=list)


class ActivityHeatmapItem(BaseModel):
    date: str
    minutes: int = 0
    events_count: int = 0


class MasteryTierTopic(BaseModel):
    id: str
    title: str
    mastery_score: float
    status: str


class MasteryDistributionResponse(BaseModel):
    mastered_count: int = 0       # score >= 0.8
    proficient_count: int = 0     # 0.5 <= score < 0.8
    needs_practice_count: int = 0 # 0.0 < score < 0.5
    unstarted_count: int = 0      # score == 0.0
    mastered_topics: list[MasteryTierTopic] = Field(default_factory=list)
    proficient_topics: list[MasteryTierTopic] = Field(default_factory=list)
    needs_practice_topics: list[MasteryTierTopic] = Field(default_factory=list)
    unstarted_topics: list[MasteryTierTopic] = Field(default_factory=list)


class LearningVelocityPoint(BaseModel):
    date: str
    topics_completed: int = 0
    study_minutes: int = 0


class WeaknessConcept(BaseModel):
    id: str | None = None
    title: str
    subject: str = ""
    source: str  # "roadmap_node" | "memory_record"
    mastery_score: float | None = None
    reason: str

File: frontend/src/api/analytics.ts
import { apiClient, ApiEnvelope } from "./client";

export interface AnalyticsEvent {
  id: string;
  profile_id: string;
  event_type: string;
  entity_type: string;
  entity_id: string | null;
  value: number | null;
  metadata_json: string | null;
  occurred_at: string;
}

export interface ChapterProgressItem {
  chapter_title: string;
  total_topics: number;
  completed_topics: number;
  completion_percentage: number;
  average_mastery: number;
}

export interface StudyTimeBreakdown {
  roadmap_minutes: number;
  chat_minutes: number;
  quiz_minutes: number;
  notes_minutes: number;
  total_minutes: number;
}

export interface AnalyticsOverview {
  completion_percentage: number;
  total_topics: number;
  completed_topics: number;
  in_progress_topics: number;
  not_started_topics: number;
  average_mastery: number;
  total_study_minutes: number;
  active_streak_days: number;
  total_sessions: number;
  total_documents: number;
  total_notes: number;
  total_quizzes_taken: number;
  average_quiz_score: number;
  study_time_breakdown: StudyTimeBreakdown;
  chapter_progress: ChapterProgressItem[];
}

export interface ActivityHeatmapItem {
  date: string;
  minutes: number;
  events_count: number;
}

export interface MasteryTierTopic {
  id: string;
  title: string;
  mastery_score: number;
  status: string;
}

export interface MasteryDistribution {
  mastered_count: number;
  proficient_count: number;
  needs_practice_count: number;
  unstarted_count: number;
  mastered_topics: MasteryTierTopic[];
  proficient_topics: MasteryTierTopic[];
  needs_practice_topics: MasteryTierTopic[];
  unstarted_topics: MasteryTierTopic[];
}

export interface LearningVelocityPoint {
  date: string;
  topics_completed: number;
  study_minutes: number;
}

export interface WeaknessConcept {
  id?: string;
  title: string;
  subject?: string;
  source: string;
  mastery_score?: number;
  reason: string;
}

export const analyticsApi = {
  getOverview: async (profileId: string): Promise<AnalyticsOverview> => {
    const res = await apiClient.get<ApiEnvelope<AnalyticsOverview>>(
      `/profiles/${profileId}/analytics/overview`
    );
    return res.data.data;
  },

  getHeatmap: async (
    profileId: string,
    days: number = 30
  ): Promise<ActivityHeatmapItem[]> => {
    const res = await apiClient.get<ApiEnvelope<ActivityHeatmapItem[]>>(
      `/profiles/${profileId}/analytics/heatmap`,
      { params: { days } }
    );
    return res.data.data;
  },

  getMastery: async (profileId: string): Promise<MasteryDistribution> => {
    const res = await apiClient.get<ApiEnvelope<MasteryDistribution>>(
      `/profiles/${profileId}/analytics/mastery`
    );
    return res.data.data;
  },

  getVelocity: async (
    profileId: string,
    days: number = 14
  ): Promise<LearningVelocityPoint[]> => {
    const res = await apiClient.get<ApiEnvelope<LearningVelocityPoint[]>>(
      `/profiles/${profileId}/analytics/velocity`,
      { params: { days } }
    );
    return res.data.data;
  },

  getWeaknesses: async (profileId: string): Promise<WeaknessConcept[]> => {
    const res = await apiClient.get<ApiEnvelope<WeaknessConcept[]>>(
      `/profiles/${profileId}/analytics/weaknesses`
    );
    return res.data.data;
  },

  logEvent: async (
    profileId: string,
    payload: {
      event_type: string;
      entity_type?: string;
      entity_id?: string;
      value?: number;
      metadata_json?: string;
    }
  ): Promise<AnalyticsEvent> => {
    const res = await apiClient.post<ApiEnvelope<AnalyticsEvent>>(
      `/profiles/${profileId}/analytics/events`,
      payload
    );
    return res.data.data;
  },
};
